src/menu.cc

```diff
@@ -105,106 +105,181 @@ extern "C" int nm_menu_hook(void *libnickel, nm_menu_item_t **items, size_t item
     #undef NM_ERR_RET
 }
 
+// AbstractNickelMenuController_createAction_before wraps
+// AbstractNickelMenuController::createAction to use the correct separator for
+// the menu location and to match the behaviour of QMenu::insertAction instead
+// of QMenu::addAction.
+QAction *AbstractNickelMenuController_createAction_before(QAction *before, nm_menu_location_t loc, bool last_in_group, void *_this, QMenu *menu, QWidget *widget, bool close, bool enabled, bool separator);
+
+// nm_menu_item_do runs a nm_menu_item_t and must be called from the thread of a
+// signal handler.
+void nm_menu_item_do(nm_menu_item_t *it);
+
+// _nm_menu_inject handles the QMenu::aboutToShow signal and injects menu items.
+void _nm_menu_inject(void *nmc, QMenu *menu, nm_menu_location_t loc, int at);
+
 extern "C" MenuTextItem* _nm_menu_hook(void* _this, QMenu* menu, QString const& label, bool checkable, bool checked, QString const& thingy) {
     NM_LOG("AbstractNickelMenuController::createMenuTextItem(%p, `%s`, %d, %d, `%s`)", menu, qPrintable(label), checkable, checked, qPrintable(thingy));
 
     QString trmm = QCoreApplication::translate("StatusBarMenuController", "Settings");
     QString trrm = QCoreApplication::translate("DictionaryActionProxy", "Dictionary");
     NM_LOG("Comparing against '%s', '%s'", qPrintable(trmm), qPrintable(trrm));
 
-    bool ismm, isrm; // only one will be true
-    if ((ismm = (label == trmm) && !checkable))
+    nm_menu_location_t loc = {};
+    if (label == trmm && !checkable) {
         NM_LOG("Intercepting main menu (label=Settings, checkable=false)...");
-    if ((isrm = (label == trrm) && !checkable))
+        loc = NM_MENU_LOCATION_MAIN_MENU;
+    } else if (label == trrm && !checkable) {
         NM_LOG("Intercepting reader menu (label=Dictionary, checkable=false)...");
+        loc = NM_MENU_LOCATION_READER_MENU;
+    }
+
+    if (loc)
+        QObject::connect(menu, &QMenu::aboutToShow, std::bind(_nm_menu_inject, _this, menu, loc, menu->actions().count()));
+
+    return AbstractNickelMenuController_createMenuTextItem_orig(_this, menu, label, checkable, checked, thingy);
+}
+
+void _nm_menu_inject(void *nmc, QMenu *menu, nm_menu_location_t loc, int at) {
+    NM_LOG("inject %d @ %d", loc, at);
+
+    NM_LOG("getting insertion point");
+
+    auto actions = menu->actions();
+    auto before = at < actions.count()
+        ? actions.at(at)
+        : nullptr;
+
+    if (before == nullptr)
+        NM_LOG("it seems the original item to add new ones before was never actually added to the menu (number of items when the action was created is %d, current is %d), appending to end instead", at, actions.count());
+
+    NM_LOG("checking for old items");
+
+    for (auto action : actions) {
+        if (action->property("nm_action") == true) {
+            return; // already added
+            /*menu->removeAction(action);
+            delete action;*/
```

Comment:
This is what we would use for #42.

Reply:
Basically, I'd uncomment these two lines and execute them (instead of returning) if the config file has changed. Then, I'd need to have something running in the background to re-parse the config file (and run generators) whenever it changes (I think the best way to keep track would be an ever-increasing counter on it). It wouldn't be too much work to extend this to also run generators periodically (it's already idempotent).

Reply:
This works because this PR allows us to rebuild any of the menus just before they are shown.

Reply:
Now that I think about it, I think I'd rather check the config file for changes (based on the mtime) before opening the menu rather than periodically. It's less likely to interfere with other stuff, a bug is less likely to cause hard-to-debug crashes or periodic hangs (specifically around the generators), it wouldn't require tracking the state of the partition, NM would remain single-threaded, it would be more efficient overall, and the extra delay when opening a menu would be negligible (a few milliseconds without generators, around 50 with, and basically nonexistent if the config files don't change).